hedera-node/hedera-app/src/main/java/com/hedera/node/app/blocks/impl/streaming/BlockBufferService.java

```diff
@@ -0,0 +1,604 @@
+     */
+    public void setLatestAcknowledgedBlock(final long blockNumber) {
+        if (!isStreamingEnabled.get()) {
+            return;
+        }
+
+        final long highestBlock = highestAckedBlockNumber.updateAndGet(current -> Math.max(current, blockNumber));
+        blockStreamMetrics.setLatestAcknowledgedBlockNumber(highestBlock);
+    }
+
+    /**
+     * Gets the current block number.
+     *
+     * @return the current block number or -1 if no blocks have been opened yet
+     */
+    public long getBlockNumber() {
+        return lastProducedBlockNumber;
+    }
+
+    /**
+     * Ensures that there is enough capacity in the block buffer to permit a new block being created. If there is not
+     * enough capacity - i.e. the buffer is saturated - then this method will block until there is enough capacity.
+     */
+    public static void ensureNewBlocksPermitted() {
+        if (!isStreamingEnabled.get()) {
+            return;
+        }
+
+        final CompletableFuture<Boolean> cf = backpressureCompletableFutureRef.get();
+        if (cf != null && !cf.isDone()) {
+            try {
+                logger.error("!!! Block buffer is saturated; blocking thread until buffer is no longer saturated");
+                final long startMs = System.currentTimeMillis();
+                final boolean bufferAvailable = cf.get();
+                final long durationMs = System.currentTimeMillis() - startMs;
+                logger.warn("Thread was blocked for {}ms waiting for block buffer to free space", durationMs);
+
+                if (!bufferAvailable) {
+                    logger.warn("Block buffer still not available to accept new blocks; reentering wait...");
+                    ensureNewBlocksPermitted();
+                }
+            } catch (final InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } catch (final Exception e) {
+                logger.warn("Failed to wait for block buffer to be available", e);
+            }
+        }
+    }
+
+    /**
+     * Prunes the block buffer by removing blocks that have been acknowledged and exceeded the configured TTL. By doing
+     * this, we also inadvertently can know if buffer is "saturated" due to blocks not being acknowledged in a timely
+     * manner.
+     */
+    private @NonNull PruneResult pruneBuffer() {
+        final Duration ttl = blockBufferTtl();
+        final Instant cutoffInstant = Instant.now().minus(ttl);
+        final Iterator<BlockState> it = blockBuffer.iterator();
+        final long highestBlockAcked = highestAckedBlockNumber.get();
+        /*
+        Calculate the ideal max buffer size. This is calculated as the block buffer TTL (e.g. 5 minutes) divided by the
+        block period (e.g. 2 seconds). This gives us an ideal number of blocks in the buffer.
+         */
+        final long idealMaxBufferSize = ttl.dividedBy(blockPeriod());
```

Comment:
It is possible to set blockPeriod as zero, which would cause Division by zero error.